### PR TITLE
Add Integration Tests for analytics core read path migration

### PR DESCRIPTION
**Pull request by shrutisinghania**
This PR adds integration tests for the Analytics Core read path migration and parameterizes existing tests for both enabled and disabled configurations. It also refines seek and range validation logic to handle edge cases like unknown file sizes and improves telemetry by tracking async read durations and GCS request statistics.

### CODE REVIEW

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamAnalyticsIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+    }
+  }
+
+  @Test
+  public void seek_beyondFileSize_throwsEOFException() throws Exception {
+    String testContent = "test content";
+    TestFixture fixture = setupFixture("seek_beyond_file", testContent);
+
+    FileSystem.Statistics statistics = new FileSystem.Statistics(fixture.ghfs.getScheme());
+    try (GoogleHadoopFSInputStream in =
```
Comment (dheerajsngh) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamAnalyticsIntegrationTest.java]: Can we use the helper method `createGhfsInputStream` at all the places?

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamAnalyticsIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    Configuration config = GoogleHadoopFileSystemIntegrationHelper.getTestConfig();
+    config.setBoolean("fs.gs.analytics.core.enable", true);
+
+    GoogleHadoopFileSystem ghfs = GoogleHadoopFileSystemIntegrationHelper.createGhfs(path, config);
+    gcsFsIHelper.writeTextFile(path, content);
+    return new TestFixture(ghfs, path, content);
+  }
+
+  @Test
+  public void read_singleBytes_returnsData() throws Exception {
```
Comment (dheerajsngh) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamAnalyticsIntegrationTest.java:71]: Currently we have test for adding single byte.
* Can we also add test to reading larger chunk `read_buffer_returnsData`
* Also a test for the case where the input buffer is already full.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamAnalyticsIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    gcsFsIHelper = GoogleCloudStorageFileSystemIntegrationHelper.create();
+    gcsFsIHelper.beforeAllTests();
+  }
+
+  @AfterClass
+  public static void afterClass() {
+    gcsFsIHelper.afterAllTests();
+  }
+
+  private static class TestFixture {
```
Comment (dheerajsngh) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamAnalyticsIntegrationTest.java]: We can move this into `GoogleHadoopFileSystemIntegrationHelper` class.

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java [hunk 4]
```diff
@@ -70,7 +69,19 @@ public long position() throws IOException {
   @Override
   public SeekableByteChannel position(long newPosition) throws IOException {
     checkOpen();
-    checkArgument(newPosition >= 0 && newPosition <= size, "Invalid position: %s", newPosition);
+    if (newPosition < 0) {
```
Comment (dheerajsngh) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java:77]: This is an additional change apart from the integration tests, lets update the PR description for it.

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java [hunk 5]
```diff
@@ -70,7 +69,19 @@ public long position() throws IOException {
...
+      GoogleCloudStorageEventBus.postOnException();
```
Comment (dheerajsngh) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java:78]: Similar to this,we should also post the exception in the GCSEventBus, at other places like `checkOpen`, all the reads, `close`. 

I guess those will be added when we will add all the logging.In that case we should have TO DO added for them.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamAnalyticsIntegrationTest.java
Comment (dheerajsngh) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamAnalyticsIntegrationTest.java:1]: Simiar comment for the test structure as previous PR. 
Follow Arrange, Act and Assert and seperate them with blank lines.

**Review by dheerajsngh — CHANGES_REQUESTED** [review]
Overall Looks Good.
Added comments for adding some more tests. 

Also can we add additional test in `GoogleHadoopFSInputStreamIntegrationTest` to tests the behavior with and without the analytics-core flag.

**Comment by dheerajsngh** [comment]
> Overall Looks Good. Added comments for adding some more tests.
> 
> Also can we add additional test in `GoogleHadoopFSInputStreamIntegrationTest` to tests the behavior with and without the analytics-core flag.

Looks like we are updating GoogleHadoopFSInputStreamIntegrationTest in the follow up PR : https://github.com/GoogleCloudDataproc/hadoop-connectors/pull/1701.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamAnalyticsIntegrationTest.java (hunk 1 above)
Comment (shrutisinghania) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamAnalyticsIntegrationTest.java]: Used `createGhfsInputStream` at all the applicable places

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamAnalyticsIntegrationTest.java (hunk 2 above)
Comment (shrutisinghania) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamAnalyticsIntegrationTest.java:71]: Added cases for these scenarios as well

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamAnalyticsIntegrationTest.java (hunk 3 above)
Comment (shrutisinghania) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamAnalyticsIntegrationTest.java]: Done

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java (hunk 4 above)
Comment (shrutisinghania) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java:77]: Updated the PR description

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java (hunk 5 above)
Comment (shrutisinghania) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java:78]: Added a TODO

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamAnalyticsIntegrationTest.java
Comment (shrutisinghania) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamAnalyticsIntegrationTest.java:1]: Done

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java [hunk 6]
```diff
@@ -31,6 +31,9 @@
 /**
  * Adapter to expose {@link GoogleCloudStorageInputStream} from Analytics Core library as {@link
  * SeekableByteChannel} for use in {@link GoogleHadoopFSInputStream}.
+ *
+ * <p>TODO(user): Wrap exception-throwing calls to post exceptions to GoogleCloudStorageEventBus on
```
Comment (dheerajsngh) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java:36]: nit: add username

**Review by dheerajsngh — APPROVED** [review]

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java [hunk 7]
```diff
@@ -70,7 +73,19 @@ public long position() throws IOException {
   @Override
   public SeekableByteChannel position(long newPosition) throws IOException {
     checkOpen();
-    checkArgument(newPosition >= 0 && newPosition <= size, "Invalid position: %s", newPosition);
+    if (newPosition < 0) {
+      GoogleCloudStorageEventBus.postOnException();
+      throw new EOFException(
+          "Invalid seek offset: position value (" + newPosition + ") must be >= 0");
+    }
+    if (size >= 0 && newPosition >= size) {
```
Comment (animesh-g) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java:85]: In previous state, newPosition == size was a valid case but here we are throwing exception. Is it intentional?

**Review by animesh-g — CHANGES_REQUESTED** [review]
I was reviewing this and realized there are redundant tests in GoogleHadoopFSInputStreamAnalyticsIntegrationTest . I think https://github.com/GoogleCloudDataproc/hadoop-connectors/pull/1701 is the right way to handle this. Let's cancel one of PR and merge changes into one for easier review and maintenance.

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java (hunk 7 above)
Comment (shrutisinghania) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java:85]: Yes, this change is intentional. It was modified to align with the behavior of existing implementations like [GoogleCloudStorageReadChannel](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/master/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageReadChannel.java#L695) and [GoogleCloudStorageClientReadChannel](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/master/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientReadChannel.java#L755).

**Comment by shrutisinghania** [comment]
> I was reviewing this and realized there are redundant tests in GoogleHadoopFSInputStreamAnalyticsIntegrationTest . I think #1701 is the right way to handle this. Let's cancel one of PR and merge changes into one for easier review and maintenance.

Integrated changes from PR #1701 into this PR

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java [hunk 8]
```diff
@@ -221,17 +220,120 @@ private GoogleHadoopFSInputStream(
...
+                    GcsObjectRange.builder()
+                        .setLength(range.getLength())
+                        .setOffset(range.getOffset())
+                        .setByteBufferFuture(range.getData())
+                        .build())
+            .collect(Collectors.toList());
+    analyticsChannel.readVectored(gcsObjectRanges, allocate);
+    // Update metrics manually as Analytics Core bypasses the standard GCS event
+    // bus.
+    storageStatistics.incrementGcsTotalRequestCount();
```
Comment (animesh-g) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java]: The Analytics Core library uses intelligent buffering, prefetching and range merging etc. Incrementing the request count on every read() call will lead to inaccurate metrics, reporting multiple GCS requests for a single actual request. I think it should be reported by analytics-core itself.
Applies to the other places in the file as well.

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java (hunk 7 above)
Comment (animesh-g) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java:85]: I am fine with the current implementation but [HDFS implementation](https://github.com/apache/hadoop/blob/da7a26b39de4791d0d73324d8ee1ff6034df9cdd/hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DFSInputStream.java#L1651) doesn't throw EOF exception on targetPos==blockEnd. However the above implementation is aligned with [Hadoop spec](https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-common/filesystem/fsdatainputstream.html#Seekable.seek.28s.29) itself. Can you add a comment with the rationale here to avoid confusion in future?

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java [hunk 9]
```diff
@@ -221,17 +220,120 @@ private GoogleHadoopFSInputStream(
...
+    storageStatistics.incrementCounter(GoogleCloudStorageStatistics.GCS_GET_MEDIA_REQUEST, 1);
```
Comment (animesh-g) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java]: nit: I think it is recording metric regardless of success or failure. Please confirm if metrics `GCS_GET_MEDIA_REQUEST` expects the same behavior.

**Review by animesh-g — COMMENTED** [review]
Integration suite looks good. I had a couple of concerns with telemetry.

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java (hunk 8 above)
Comment (shrutisinghania) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java]: As discussed offline, removed the logic to manually update metrics and ignored the tests for metric check

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java (hunk 9 above)
Comment (shrutisinghania) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java]: Removed this metric

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java (hunk 7 above)
Comment (shrutisinghania) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GcsAnalyticsCoreInputStreamWrapper.java:85]: Added a comment

**Review by animesh-g — APPROVED** [review]

**Review by animesh-g — APPROVED** [review]
